### PR TITLE
Boehm GC Fixes - SRE and Handle Stack

### PR DESCRIPTION
**Pull request by joncham**
The contained GC tracked hash tables are now themselves roots since https://github.com/mono/mono/pull/4248

Fixes crash due to MonoDynamicAssembly or MonoDynamicImage being collected under Boehm as they were unreferenced GC memory.

### CODE REVIEW

File: mono/metadata/sre.c [hunk 1]
```diff
@@ -1222,12 +1222,7 @@ mono_reflection_dynimage_basic_init (MonoReflectionAssemblyBuilder *assemblyb)
 	if (assemblyb->dynamic_assembly)
 		return;
 
-#if HAVE_BOEHM_GC
-	/* assembly->assembly.image might be GC allocated */
-	assembly = assemblyb->dynamic_assembly = (MonoDynamicAssembly *)GC_MALLOC (sizeof (MonoDynamicAssembly));
```
Comment (joncham) [inline mono/metadata/sre.c]: I cannot find where we free the `MonoDynamicAssembly` we allocate?

**Comment by vargaz** [comment]
Looks good, one less #ifdef BOEHM.

**Comment by joncham** [comment]
@vargaz did you see comment about freeing the allocated `MonoDynamicAssembly`? Do we just leak that currently?

**Comment by vargaz** [comment]
Probably.

**Comment by joncham** [comment]
Switched to functions rather than macros.

**Comment by lambdageek** [comment]
build

File: mono/metadata/handle.c [hunk 2]
```diff
@@ -245,9 +294,9 @@ mono_handle_new_interior (gpointer rawptr, const char *owner)
 HandleStack*
 mono_handle_stack_alloc (void)
 {
-	HandleStack *stack = g_new0 (HandleStack, 1);
-	HandleChunk *chunk = g_new0 (HandleChunk, 1);
-	HandleChunk *interior = g_new0 (HandleChunk, 1);
+	HandleStack *stack = new_handle_stack ();
+	HandleChunk *chunk = new_handle_chunk ();
+	HandleChunk *interior = new_handle_chunk ();
```
Comment (lambdageek) [inline mono/metadata/handle.c]: Oops... just realized that this isn't okay actually... we were relying on `g_new0` to initialize `chunk->size`, `chunk->next`, `chunk->prev` and `interior->size` to 0.

**Comment by lambdageek** [comment]
Failures are unrelated

**Pull request merged**